twitter-as.js

```diff
@@ -0,0 +1,232 @@
+var RemoteRoom = require("matrix-appservice-bridge").RemoteRoom;
+var RemoteRoom = require("matrix-appservice-bridge").MatrixRoom;
+var AppServiceRegistration = require("matrix-appservice-bridge").AppServiceRegistration;
+var Twitter;
+
+function roomPowers(users) {
+    return {
+        "content": {
+            "ban": 50,
+            "events": {
+                "m.room.name": 100,
+                "m.room.power_levels": 100,
+                "m.room.topic": 100,
+                "m.room.join_rules": 100,
+                "m.room.avatar": 100,
+                "m.room.aliases": 75,
+                "m.room.canonical_alias": 75
+            },
+            "events_default": 10,
+            "kick": 75,
+            "redact": 75,
+            "state_default": 0,
+            "users": users,
+            "users_default": 0
+        },
+        "state_key": "",
+        "type": "m.room.power_levels"
+    };
+}
+
+new Cli({
+    registrationPath: "twitter-registration.yaml",
+    bridgeConfig: {
+        schema: "config.yaml",
+        defaults: {
+            test: "ABC"
+        }
+    },
+    generateRegistration: function(reg, callback) {
+        reg.setId(AppServiceRegistration.generateToken());
+        reg.setHomeserverToken(AppServiceRegistration.generateToken());
+        reg.setAppServiceToken(AppServiceRegistration.generateToken());
+        reg.setSenderLocalpart("twitbot");
+        reg.addRegexPattern("users", "@twitter_.*", true);
+        reg.addRegexPattern("aliases", "#twitter_@.*", true);
+        callback(reg);
+    },
+    run: function(port, config) {
+        bridge = new Bridge({
+            homeserverUrl: "http://localhost:8008",
+            domain: "localhost",
+            registration: "twitter-registration.yaml",
+            controller: {
+                onUserQuery: function(queriedUser) {
+                    var twitter_user;
+                    var image_uri;
+                    return Twitter.get_user_by_id(queriedUser.localpart.substr(8)).then( (tuser) => {
+                      twitter_user = tuser;
+                      return uploadImageFromUrl(bridge, twitter_user.profile_image_url_https, queriedUser.getId()).then((uri) => {
+                          console.log("Got User Avatar:", uri);
+                          image_uri = uri;
+                      });
+                    }).then(() => {
+                      var remoteUser = new RemoteUser(twitter_user.id_str);
+                      var user = {
+                          name: twitter_user.name + " (@" + twitter_user.screen_name + ")",
+                          url: image_uri,
+                          remote: remoteUser
+                      };
+                      console.log(user);
+                      return user;
+                    }).catch((error) => {
+                        console.error("Couldn't find the user.");
+                        console.error("Reason:", error);
+                    });
```

Comment:
This is a gnarly promise chain which isn't really passing anything down the chain and instead setting stuff on the in-scope variables `twitter_user` and `image_uri`. You can do this better:

```javascript
onUserQuery: function(queriedUser) {
  return Twitter.get_user_by_id(queriedUser.localpart.substr("twitter_".length)).then( (twitter_user) => {
    return uploadImageFromUrl(bridge, twitter_user.profile_image_url_https, queriedUser.getId()).then((uri) => {
      return {
        name: twitter_user.name + " (@" + twitter_user.screen_name + ")",
        url: uri,
        remote: new RemoteUser(twitter_user.id_str)
      };
    );
  }).catch((err) => {
    console.error("Failed to query user: ", err);
  });
}
```

Alternatively, say goodbye to `Promise` and use something like Bluebird's coroutines which make reading this a lot easier - see https://github.com/matrix-org/matrix-appservice-irc/blob/develop/lib/bridge/IrcBridge.js#L269 for an example (look for `yield`).